File: lib/domain/mapper/weather_data_mapper.dart
import 'package:logger/logger.dart';

import 'package:avocado/domain/model/weather_model.dart';

Weather getUltraShortTermLiveMapper(Map<String, dynamic> data) {
  List items = data['response']['body']['items']['item'];

  Map<String, dynamic> temp = {};

  try {
    for (var item in items) {
      String? category = item['category'];

      if (category != null) {
        String? parsedCategory = _category[category];

        if (parsedCategory != null) {
          String? value = item['obsrValue'] ?? '';
          temp[parsedCategory] = value;
        } else {
          continue;
        }
      } else {
        continue;
      }
    }
  } catch (e) {
    Logger().e(e);
  }

  Weather weather = Weather.fromJson(temp);

  return weather;
}

//NOTE *****************************************************
// 하늘상태(SKY) 코드 : 맑음(1), 구름많음(3), 흐림(4)
// 강수형태(PTY) 코드 : 없음(0), 비(1), 비/눈(2), 눈(3), 빗방울(5), 빗방울눈날림(6), 눈날림(7)
// *****************************************************
WeatherType getWeatherType(Map<String, String> data) {
  try {
    String sky = data['sky']!;
    String precipitationType = data['precipitationType']!;
    String precipitationPerHour = data['precipitationPerHour']!;
    String lightning = data['lightning']!;

    WeatherType weatherType = WeatherType.sunny;

    switch (precipitationType) {
      case '0':
        switch (sky) {
          case '1':
            weatherType = WeatherType.sunny;
          case '3':
            weatherType = WeatherType.cloudyPartly;
          case '4':
            weatherType = WeatherType.cloudyNormal;
        }
        break;
      case '1':
        String preprecipitationPerHourWithoutUnit =
            precipitationPerHour.split('mm')[0];
        double parsedPrecipitationPerHour =
            double.parse(preprecipitationPerHourWithoutUnit);

        //NOTE - 세계기상기구 기준, 비 강도
        if (parsedPrecipitationPerHour < 3) {
          weatherType = WeatherType.rainningDrizzle;
        } else if (parsedPrecipitationPerHour < 15) {
          weatherType = WeatherType.rainningNormal;
        } else if (parsedPrecipitationPerHour < 30) {
          weatherType = WeatherType.rainningHeavily;
        } else {
          weatherType = WeatherType.rainningDownpour;
        }
        break;
      case '2':
        //TODO - 비/눈, API 설명이 부족함
        break;
      case '3':
        weatherType = WeatherType.snowing;
        break;
      case '5':
        //TODO - 빗방울, API 설명이 부족함
        break;
      case '6':
        //TODO - 빗방울눈날림, API 설명이 부족함
        break;
      case '7':
        //TODO - 눈날림, API 설명이 부족함
        break;
      default:
    }

    return weatherType;
  } catch (e) {
    Logger().e(e);
    return WeatherType.initial;
  }
}

Weather getUltraShortTermForecastMapper(Map<String, dynamic> data) {
  if (data['response'] == null ||
      data['response']['body'] == null ||
      data['response']['body']['items'] == null ||
      data['response']['body']['items']['item'] == null) {
    throw '데이터가 없습니다.';
  }

  List items = data['response']['body']['items']['item'];

  Map<String, List> parsedItems = {};

  // INFO: 시간대별 데이터 파싱
  for (var item in items) {
    if (parsedItems[item['fcstTime']] == null) {
      parsedItems[item['fcstTime']] = [item];
    } else {
      parsedItems[item['fcstTime']]!.add(item);
    }
  }

  // INFO: 현재 시간 아이템 필터
  var filtered =
      parsedItems['${DateTime.now().hour.toString().padLeft(2, '0')}00'];

  Map<String, String> parsedByCategory = {};

  // INFO: 카테고리별 value 값 할당
  try {
    for (var item in filtered!) {
      String? category = item['category'];

      if (category != null) {
        String? parsedCategory = _category[category];

        if (parsedCategory != null) {
          String value = item['fcstValue'] ?? '';
          parsedByCategory[parsedCategory] = value;
        } else {
          continue;
        }
      } else {
        continue;
      }
    }
  } catch (e) {
    Logger().e(e);
  }

  Logger().d(parsedByCategory);

  WeatherType weatherType = getWeatherType(parsedByCategory);

  return Weather.fromJson(parsedByCategory).copyWith(type: weatherType);
}

Map<String, String> _category = {
  'T1H': 'temperature',
  'RN1': 'precipitationPerHour',
  'SKY': 'sky',
  'UUU': 'windSpeedToEastWest',
  'VVV': 'windSpeedToSouthNorth',
  'REH': 'huminity',
  'PTY': 'precipitationType',
  'LGT': 'lightning',
  'VEC': 'windDirection',
  'WSD': 'windSpeed',
};

File: functions/lib/utils/weatherDataMapper.ts
import { logger } from 'firebase-functions/v1';

interface ForecastResponse {
  response: {
    body: {
      items: {
        item: Array<{
          fcstTime: string;
          category: string;
          fcstValue: string;
        }>;
      };
    };
  };
}

export interface UltraShortForecastWeather {
  temperature?: string;
  precipitationPerHour?: string;
  sky?: string;
  windSpeedToEastWest?: string;
  windSpeedToSouthNorth?: string;
  huminity?: string;
  precipitationType?: string;
  lightning?: string;
  windDirection?: string;
  windSpeed?: string;
  type?: string | WeatherType;
}

export interface ShortForecastWeather {
  precipitationProbability?: string;
  precipitationType?: string;
  precipitationPerHour?: string;
  huminity?: string;
  snowPerHour?: string;
  sky?: string;
  temperaturePerHour?: string;
  temperatureMin?: string;
  temperatureMax?: string;
  windSpeedToEastWest?: string;
  windSpeedToSouthNorth?: string;
  wave?: string;
  windDirection?: string;
  windSpeed?: string;
  type?: string | WeatherType;
  time: string;
}

export enum WeatherType {
    initial = 'initial',
    rainningDrizzle = 'rainningDrizzle',
    rainningNormal = 'rainningNormal',
    rainningHeavily = 'rainningHeavily',
    rainningDownpour = 'rainningDownpour',
    cloudyPartly = 'cloudyPartly',
    cloudyNormal = 'cloudyNormal',
    sunny = 'sunny',
    snowing = 'snowing'
}
  
function getWeatherType(data: UltraShortForecastWeather | ShortForecastWeather): WeatherType {
try {
    const sky = data['sky']!;
    const precipitationType = data['precipitationType']!;
    const precipitationPerHour = data['precipitationPerHour']!;

    let weatherType: WeatherType = WeatherType.sunny;

    switch (precipitationType) {
    case '0':
        switch (sky) {
        case '1':
            weatherType = WeatherType.sunny;
            break;
        case '3':
            weatherType = WeatherType.cloudyPartly;
            break;
        case '4':
            weatherType = WeatherType.cloudyNormal;
            break;
        }
        break;
    case '1':
        const preprecipitationPerHourWithoutUnit = precipitationPerHour.split('mm')[0];
        const parsedPrecipitationPerHour = parseFloat(preprecipitationPerHourWithoutUnit);

        // NOTE - 세계기상기구 기준, 비 강도
        if (parsedPrecipitationPerHour < 3) {
            weatherType = WeatherType.rainningDrizzle;
        } else if (parsedPrecipitationPerHour < 15) {
            weatherType = WeatherType.rainningNormal;
        } else if (parsedPrecipitationPerHour < 30) {
            weatherType = WeatherType.rainningHeavily;
        } else {
            weatherType = WeatherType.rainningDownpour;
        }
        break;
    case '2':
        // TODO - 비/눈, API 설명이 부족함
        break;
    case '3':
        weatherType = WeatherType.snowing;
        break;
    case '5':
        // TODO - 빗방울, API 설명이 부족함
        break;
    case '6':
        // TODO - 빗방울눈날림, API 설명이 부족함
        break;
    case '7':
        // TODO - 눈날림, API 설명이 부족함
        break;
    default:
        break;
    }

    return weatherType;
} catch (e) {
    console.error(e);
    return WeatherType.initial;
}
}


function sortByForecastTime(data: Array<{
  fcstTime: string;
  category: string;
  fcstValue: string;
}>) {
  const parsedData: Record<string, Array<{ fcstTime: string; category: string; fcstValue: string }>> = {};

  // Parse the time-based data
  for (const item of data) {
    if (!parsedData[item.fcstTime]) {
      parsedData[item.fcstTime] = [item];
    } else {
      parsedData[item.fcstTime]!.push(item);
    }
  }

  return parsedData;
}

export function getUltraShortTermForecastMapper(data: ForecastResponse): UltraShortForecastWeather | null {
  try {

    if (!data.response || !data.response.body || !data.response.body.items || !data.response.body.items.item) {
      throw new Error('No data available');
    }

    const items = data.response.body.items.item;
    
    const parsedItems = sortByForecastTime(items);

    const currentTimeItems = parsedItems[`${String(new Date().getHours()).padStart(2, '0')}00`];

    const parsedByCategory: UltraShortForecastWeather = {};

    for (const item of currentTimeItems!) {
      const category = item.category;

      if (category && ultraShortTermForecastCategoryMap[category]) {
        const parsedCategory = ultraShortTermForecastCategoryMap[category];
        parsedByCategory[parsedCategory] = item.fcstValue || '';
      }
    }
 

    const weatherType = getWeatherType(parsedByCategory);

    return {
      ...parsedByCategory,
      type: weatherType,
    };
  } catch (e) {
    logger.error(e);
    return null;
  }
}

export type GetShortTermForecastMapperPayload = {
  data:ShortForecastWeather[] | null,
  rainningCountOnDistrict: number,
} | null;

export function getShortTermForecastMapper(data: ForecastResponse): GetShortTermForecastMapperPayload {
  try {

    if (!data.response || !data.response.body || !data.response.body.items || !data.response.body.items.item) {
      throw new Error('No data available');
    }

    const items = data.response.body.items.item;
    
    const parsedItems = sortByForecastTime(items);

    
    const times = Array(4).fill(new Date().getHours()).map(function(value, index){
      let hour = value + (index * 2);

      if(hour >= 24){
        hour -= 24;
      }

      return `${String(hour).padStart(2, '0')}00`;
    } );

    let selectedItems: ShortForecastWeather[] = [];

    for(const time of times) {
      
      if(!!parsedItems[time]){
      
        const parsedByCategory: ShortForecastWeather = {
          time,
        };

        for (const item of parsedItems[time]) {
          const category = item.category;
       
          if (category && shortTermForecastCategoryMap[category]) {
            const parsedCategory = shortTermForecastCategoryMap[category];
            parsedByCategory[parsedCategory] = item.fcstValue || '';
          }
        }

        selectedItems.push(parsedByCategory);
      }
      
    }

    let rainningCountOnDistrict = 0;

    selectedItems = selectedItems.map((item) => {
      const type = getWeatherType(item);
      
      if(getIsRainning(type)) {
        rainningCountOnDistrict++;
      }

      return {
        ...item,
        type,
      };
    });
    
    return {
      data: selectedItems,
      rainningCountOnDistrict,
    };
  } catch (e) {
    logger.error(e);
    return null;
  }
}

const ultraShortTermForecastCategoryMap: Record<string, keyof UltraShortForecastWeather> = {
  T1H: 'temperature',
  RN1: 'precipitationPerHour',
  SKY: 'sky',
  UUU: 'windSpeedToEastWest',
  VVV: 'windSpeedToSouthNorth',
  REH: 'huminity',
  PTY: 'precipitationType',
  LGT: 'lightning',
  VEC: 'windDirection',
  WSD: 'windSpeed',
};

const shortTermForecastCategoryMap: Record<string, keyof ShortForecastWeather> = {
  POP: 'precipitationProbability',
  PTY: 'precipitationType',
  PCP: 'precipitationPerHour',
  REH: 'huminity',
  SNO: 'snowPerHour',
  SKY: 'sky',
  TMP: 'temperaturePerHour',
  TMN: 'temperatureMin',
  TMX: 'temperatureMax',
  UUU: 'windSpeedToEastWest',
  VVV: 'windSpeedToSouthNorth',
  WAV: 'wave',
  VEC: 'windDirection',
  WSD: 'windSpeed',
}

export function getIsRainning(type: WeatherType) {
  return type === WeatherType.rainningDownpour || WeatherType.rainningDrizzle || WeatherType.rainningHeavily || WeatherType.rainningNormal
}